velero-csi-plugin/util.go

```diff
@@ -0,0 +1,143 @@
+package main
+
+import (
+	"strings"
+
+	snapshotv1beta1api "github.com/kubernetes-csi/external-snapshotter/v2/pkg/apis/volumesnapshot/v1beta1"
+	snapshotter "github.com/kubernetes-csi/external-snapshotter/v2/pkg/client/clientset/versioned/typed/volumesnapshot/v1beta1"
+	"github.com/pkg/errors"
+	corev1api "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	corev1client "k8s.io/client-go/kubernetes/typed/core/v1"
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+const (
+	//TODO: use annotation from velero https://github.com/vmware-tanzu/velero/pull/2283
+	resticPodAnnotation = "backup.velero.io/backup-volumes"
+)
+
+func getPVForPVC(pvc *corev1api.PersistentVolumeClaim, corev1 corev1client.PersistentVolumesGetter) (*corev1api.PersistentVolume, error) {
+	if pvc.Spec.VolumeName == "" {
+		return nil, errors.Errorf("PVC %s/%s has no volume backing this claim", pvc.Namespace, pvc.Name)
+	}
+	if pvc.Status.Phase != corev1api.ClaimBound {
+		// TODO: confirm if this PVC should be snapshotted if it has no PV bound
+		return nil, errors.Errorf("PVC %s/%s is in phase %v and is not bound to a volume", pvc.Namespace, pvc.Name, pvc.Status.Phase)
+	}
+	pvName := pvc.Spec.VolumeName
+	pv, err := corev1.PersistentVolumes().Get(pvName, metav1.GetOptions{})
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get PV %s for PVC %s/%s", pvName, pvc.Namespace, pvc.Name)
+	}
+	return pv, nil
+}
+
+func getPodsUsingPVC(pvcNamespace, pvcName string, corev1 corev1client.PodsGetter) ([]corev1api.Pod, error) {
+	podsUsingPVC := []corev1api.Pod{}
+	for {
+		podList, err := corev1.Pods(pvcNamespace).List(metav1.ListOptions{})
+		if err != nil {
+			return podsUsingPVC, err
+		}
+
+		for _, p := range podList.Items {
+			for _, v := range p.Spec.Volumes {
+				if v.PersistentVolumeClaim != nil && v.PersistentVolumeClaim.ClaimName == pvcName {
+					podsUsingPVC = append(podsUsingPVC, p)
+				}
+			}
+		}
+		if podList.Continue == "" {
```

Comment:
this might be a little confusing given that we're not actually using a limit on the List call - let's simplify to just do a single list and iterate through it, rather than having the outer loop.

Reply:
I had understood the `List` API incorrectly.
Yeah removing the outer loop makes sense,